### DISCUSSION TITLE
Incorrect translation context in Sketcher context menu

### DISCUSSION THREAD

**Discussion by chennes** [FreeCAD/FreeCAD#9584] [Code contribution discussions]
I am working on addressing https://github.com/FreeCAD/FreeCAD-translations/issues/219 -- an entire context menu that is basically untranslated. To that end, In TaskSketcherElements.cpp, we have a function...
https://github.com/FreeCAD/FreeCAD/blob/063df7044d45026162d8ae6e21bdaaa6e1d4c858/src/Mod/Sketcher/Gui/TaskSketcherElements.cpp#L459-L520
That `CONTEXT_ITEM` macro is defined here:
https://github.com/FreeCAD/FreeCAD/blob/063df7044d45026162d8ae6e21bdaaa6e1d4c858/src/Mod/Sketcher/Gui/TaskSketcherElements.cpp#L65-L69

The macro calls `tr()` on the menu text, which won't work because the context of each of those commands, when originally created, is defined as the name of the command class itself (e.g. for `Constraint_PointOnPoint` it is `CmdSketcherConstrainCoincident`, etc.). And because the string is being passed through a macro, Qt's translation extraction won't see it as a new string, so the string MUST already be getting translated elsewhere.

There are a few possible solutions, none of them perfectly pleasant:
1. Add another parameter to the macro to give the translation context -- the macro is already complex, though, and this will make it worse. Plus, it's a macro in modern C++ code, so the real answer is probably to refactor to get rid of it. Which is a pretty big change this close to a release.
2. Do the translation at the call site instead of in the macro. Makes the call rather long and ugly.
3. Add a `#if 0` block with all the needed `QT_TRANSLATE_NOOP` calls for these strings. Requires duplicating all of the strings, and doubles the translation work (each string will end up getting translated both here and for the CommandConstraints.cpp file).

@wwmayer ,@abdullahtahiriyo, it looks like this is mostly your code. How would you prefer I handle it?

**Comment by wwmayer** [comment]
I would go for option 3. After the release I think it's best to refactor the code and try to get rid of the macro.

**Comment by abdullahtahiriyo** [comment]
3 sounds best to me. 

That file will see a big refactor during the next development cycle.

**Comment by chennes** [comment]
Sounds good, I'll code that up.
